internal/envoy/v3/listener.go

```diff
@@ -257,23 +257,43 @@ func (b *httpConnectionManagerBuilder) DefaultFilters() *httpConnectionManagerBu
 }
 
 // AddFilter appends f to the list of filters for this HTTPConnectionManager. f
-// may by nil, in which case it is ignored.
+// may by nil, in which case it is ignored. Note that Router filters
+// (filters with TypeUrl `type.googleapis.com/envoy.extensions.filters.http.router.v3.Router`)
+// are specially treated. There may only be one of these filters, and it must be the last.
+// AddFilter will ensure that the router filter, if present, is last, and will panic
+// if a second Router is added when one is already present.
 func (b *httpConnectionManagerBuilder) AddFilter(f *http.HttpFilter) *httpConnectionManagerBuilder {
 	if f == nil {
 		return b
 	}
 
-	if len(b.filters) > 0 {
-		lastIndex := len(b.filters) - 1
-		// If the router filter is last, keep it at the end
-		// of the filter chain when we add the new filter.
-		if b.filters[lastIndex].Name == wellknown.Router {
-			b.filters = append(b.filters[:lastIndex], f, b.filters[lastIndex])
-			return b
+	b.filters = append(b.filters, f)
+
+	if len(b.filters) == 1 {
+		return b
+	}
+
+	lastIndex := len(b.filters) - 1
+	routerIndex := -1
+	for i, filter := range b.filters {
+		if filter.GetTypedConfig().TypeUrl == HTTPFilterRouter {
+			routerIndex = i
 		}
 	}
 
-	b.filters = append(b.filters, f)
+	// We can't add more than one router entry, and there should be no way to do it.
+	// If this happens, it has to be programmer error, so we panic to tell them
+	// it needs to be fixed.
+	if routerIndex != -1 && f.GetTypedConfig().TypeUrl == HTTPFilterRouter {
```

Comment:
We can leave this as it is, but maybe a cleaner way to test the type is to use  [`ptypes.Is()`](https://godoc.org/github.com/golang/protobuf/ptypes#Is), .e.g.

```Go
if ptypes.Is(f.GetTypedConfig(), &Router{}) {
        // Do the thing ....
}
```

Reply:
Or we also have `protobuf.AnyMessageTypeOf()` to generate a type URL from a message struct.

Reply:
Swapped to `ptypes.Is()`.